utils/s3/client.cc

```diff
@@ -500,14 +500,24 @@ class client::upload_sink_base : public data_sink_impl {
         return !_upload_id.empty();
     }
 
-public:
-    upload_sink_base(shared_ptr<client> cln, sstring object_name, std::optional<tag> tag)
+    multipart_upload(shared_ptr<client> cln, sstring object_name, std::optional<tag> tag)
         : _client(std::move(cln))
         , _object_name(std::move(object_name))
         , _tag(std::move(tag))
     {
     }
 
+public:
+    unsigned parts_count() const noexcept { return _part_etags.size(); }
+};
+
+class client::upload_sink_base : public multipart_upload, public data_sink_impl {
```

Comment:
if `upload_sink_base` is but *using* `multipart_upload`, can we private-inherit from it? the same applies to `do_upload_file`.

Reply:
No, it complains in a tricky way. I made upload_sink_base protected inherit from multipart_upload

```
utils/s3/client.cc:786:72: error: '_object_name' is a protected member of 's3::client::multipart_upload'
  786 |     s3l.trace("PUT part {} from {} (upload id {})", part_number, piece._object_name, _upload_id);
      |                                                                        ^
```

that's the code

```c++
future<> client::multipart_upload::upload_part(std::unique_ptr<upload_sink> piece_ptr) {
    if (!upload_started()) {
        co_await start_upload();
    }

    auto& piece = *piece_ptr;
    unsigned part_number = _part_etags.size();
    _part_etags.emplace_back();
    s3l.trace("PUT part {} from {} (upload id {})", part_number, piece._object_name, _upload_id);
```

the multipart_upload tries to access field from one of its derived classes and fails :(

Reply:
But private inheriting of do_upload_file works fine